testng-core/src/main/java/org/testng/TestRunner.java

```diff
@@ -486,15 +486,13 @@ private void initMethods() {
     // Walk through all the TestClasses, store their method
     // and initialize them with the correct ITestClass
     //
+
     for (ITestClass tc : m_classMap.values()) {
       fixMethodsWithClass(tc.getTestMethods(), tc, testMethods);
-      fixMethodsWithClass(
-          ((ITestClassConfigInfo) tc).getAllBeforeClassMethods().toArray(new ITestNGMethod[0]),
-          tc,
-          beforeClassMethods);
+      fixMethodsWithClass(beforeClassConfigMethods(tc), tc, beforeClassMethods);
       fixMethodsWithClass(tc.getBeforeTestMethods(), tc, null);
       fixMethodsWithClass(tc.getAfterTestMethods(), tc, null);
-      fixMethodsWithClass(tc.getAfterClassMethods(), tc, afterClassMethods);
+      fixMethodsWithClass(afterClassConfigMethods(tc), tc, afterClassMethods);
```

Comment:
Sorry, I didn't catch it before but why not have a `fixMethodsWithClass(Array)` that will call `fixMethodsWithClass(List)`? That will be a good start for cleaning without the need to change everything.

Reply:
The change to list has been made and pushed already. Lets go ahead with that. 

I will raise a separate PR, that will deprecate all the user facing methods on `ITestNGMethod` that return an array and have it replaced with a List.